From d3ec48ab77b1c821a75ca704931494ac738cf780 Mon Sep 17 00:00:00 2001
From: Preston Cody <codeman@google.com>
Date: Sun, 14 Jun 2026 00:07:54 +0000
Subject: [PATCH] Fix out-of-bounds write in SecureElement openLogicalChannel

The NXP secure_element HAL previously failed to validate the channel number returned by the Secure Element during the MANAGE CHANNEL OPEN operation. A compromised Secure Element could return an out-of-bounds channel number, resulting in an intra-object out-of-bounds write when setting a boolean value in the `mOpenedChannels` array.

This patch addresses the vulnerability by adding boundary checks across both pn8x and snxxx HAL versions. It validates the returned channel number (`rspApdu.p_data[0]`) against the maximum allowed logical channels (`MAX_LOGICAL_CHANNELS` for pn8x and `mMaxChannelCount` for snxxx) and ensures it is greater than zero before using it as an array index. If the channel number is invalid, the code safely handles the error by treating it as `CHANNEL_NOT_AVAILABLE`. Additionally, this adds a missing early return in `closeChannel` for pn8x variants when an invalid channel number is encountered.

Bug: 522374659
Test: built and atest locally
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

Change-Id: I54a5cb16202b49d0929e6306626ef11e2aa329b8
Merged-In: I54a5cb16202b49d0929e6306626ef11e2aa329b8
(cherry picked from commit 8ca0d691fa4b34b7b60a141d5db116e153470712)
---
 pn8x/1.0/SecureElement.cpp   | 14 +++++++++-----
 pn8x/1.1/SecureElement.cpp   | 14 +++++++++-----
 snxxx/1.0/SecureElement.cpp  | 13 +++++++++----
 snxxx/1.0/VirtualISO.cpp     | 13 +++++++++----
 snxxx/1.1/SecureElement.cpp  | 13 +++++++++----
 snxxx/1.1/VirtualISO.cpp     | 13 +++++++++----
 snxxx/1.2/SecureElement.cpp  | 13 +++++++++----
 snxxx/1.2/VirtualISO.cpp     | 13 +++++++++----
 snxxx/aidl/SecureElement.cpp | 13 +++++++++----
 snxxx/aidl/VirtualISO.cpp    | 13 +++++++++----
 10 files changed, 90 insertions(+), 42 deletions(-)

diff --git a/pn8x/1.0/SecureElement.cpp b/pn8x/1.0/SecureElement.cpp
index b2159a6..10415bb 100644
--- a/pn8x/1.0/SecureElement.cpp
+++ b/pn8x/1.0/SecureElement.cpp
@@ -161,10 +161,14 @@ Return<void> SecureElement::openLogicalChannel(const hidl_vec<uint8_t>& aid,
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
     /*ManageChannel successful*/
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < MAX_LOGICAL_CHANNELS && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x6A &&
              rspApdu.p_data[rspApdu.len - 1] == 0x81) {
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
@@ -356,7 +360,7 @@ SecureElement::closeChannel(uint8_t channelNumber) {
   if ((channelNumber >= MAX_LOGICAL_CHANNELS) ||
       (mOpenedChannels[channelNumber] == false)) {
     ALOGE("%s: invalid channel!!!", __func__);
-    sestatus = SecureElementStatus::FAILED;
+    return SecureElementStatus::FAILED;
   } else if (channelNumber > DEFAULT_BASIC_CHANNEL) {
     phNxpEse_memset(&cmdApdu, 0x00, sizeof(phNxpEse_data));
     phNxpEse_memset(&rspApdu, 0x00, sizeof(phNxpEse_data));
diff --git a/pn8x/1.1/SecureElement.cpp b/pn8x/1.1/SecureElement.cpp
index c960568..01bbcbd 100644
--- a/pn8x/1.1/SecureElement.cpp
+++ b/pn8x/1.1/SecureElement.cpp
@@ -203,10 +203,14 @@ Return<void> SecureElement::openLogicalChannel(const hidl_vec<uint8_t>& aid,
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
     /*ManageChannel successful*/
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < MAX_LOGICAL_CHANNELS && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x6A &&
              rspApdu.p_data[rspApdu.len - 1] == 0x81) {
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
@@ -397,7 +401,7 @@ Return<SecureElementStatus> SecureElement::closeChannel(uint8_t channelNumber) {
   if ((channelNumber >= MAX_LOGICAL_CHANNELS) ||
       (mOpenedChannels[channelNumber] == false)) {
     ALOGE("%s: invalid channel!!!", __func__);
-    sestatus = SecureElementStatus::FAILED;
+    return SecureElementStatus::FAILED;
   } else if (channelNumber > DEFAULT_BASIC_CHANNEL) {
     phNxpEse_memset(&cmdApdu, 0x00, sizeof(phNxpEse_data));
     phNxpEse_memset(&rspApdu, 0x00, sizeof(phNxpEse_data));
diff --git a/snxxx/1.0/SecureElement.cpp b/snxxx/1.0/SecureElement.cpp
index f605957..7ba0d88 100644
--- a/snxxx/1.0/SecureElement.cpp
+++ b/snxxx/1.0/SecureElement.cpp
@@ -278,10 +278,15 @@ Return<void> SecureElement::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/1.0/VirtualISO.cpp b/snxxx/1.0/VirtualISO.cpp
index 12c2197..e8f1ce5 100644
--- a/snxxx/1.0/VirtualISO.cpp
+++ b/snxxx/1.0/VirtualISO.cpp
@@ -216,10 +216,15 @@ Return<void> VirtualISO::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/1.1/SecureElement.cpp b/snxxx/1.1/SecureElement.cpp
index 8b27e77..27e5d4e 100644
--- a/snxxx/1.1/SecureElement.cpp
+++ b/snxxx/1.1/SecureElement.cpp
@@ -364,10 +364,15 @@ Return<void> SecureElement::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/1.1/VirtualISO.cpp b/snxxx/1.1/VirtualISO.cpp
index ae90211..1ed15dc 100644
--- a/snxxx/1.1/VirtualISO.cpp
+++ b/snxxx/1.1/VirtualISO.cpp
@@ -283,10 +283,15 @@ Return<void> VirtualISO::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/1.2/SecureElement.cpp b/snxxx/1.2/SecureElement.cpp
index 8cdb5c4..ad72b9a 100644
--- a/snxxx/1.2/SecureElement.cpp
+++ b/snxxx/1.2/SecureElement.cpp
@@ -480,10 +480,15 @@ Return<void> SecureElement::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/1.2/VirtualISO.cpp b/snxxx/1.2/VirtualISO.cpp
index 445d97d..314106d 100644
--- a/snxxx/1.2/VirtualISO.cpp
+++ b/snxxx/1.2/VirtualISO.cpp
@@ -305,10 +305,15 @@ Return<void> VirtualISO::openLogicalChannel(const hidl_vec<uint8_t>& aid,
     sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SecureElementStatus::SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SecureElementStatus::SUCCESS;
+    } else {
+      resApduBuff.channelNumber = 0xff;
+      sestatus = SecureElementStatus::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/aidl/SecureElement.cpp b/snxxx/aidl/SecureElement.cpp
index ca80d6b..2306105 100644
--- a/snxxx/aidl/SecureElement.cpp
+++ b/snxxx/aidl/SecureElement.cpp
@@ -414,10 +414,15 @@ ScopedAStatus SecureElement::openLogicalChannel(
     sestatus = ISecureElement::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SESTATUS_SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SESTATUS_SUCCESS;
+    } else {
+      resApduBuff.channelNumber = -1;
+      sestatus = ISecureElement::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
diff --git a/snxxx/aidl/VirtualISO.cpp b/snxxx/aidl/VirtualISO.cpp
index 73563de..499b5df 100644
--- a/snxxx/aidl/VirtualISO.cpp
+++ b/snxxx/aidl/VirtualISO.cpp
@@ -244,10 +244,15 @@ ScopedAStatus VirtualISO::openLogicalChannel(
     sestatus = ISecureElement::CHANNEL_NOT_AVAILABLE;
   } else if (rspApdu.p_data[rspApdu.len - 2] == 0x90 &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-    resApduBuff.channelNumber = rspApdu.p_data[0];
-    mOpenedchannelCount++;
-    mOpenedChannels[resApduBuff.channelNumber] = true;
-    sestatus = SESTATUS_SUCCESS;
+    if (rspApdu.p_data[0] < mMaxChannelCount && rspApdu.p_data[0] > 0) {
+      resApduBuff.channelNumber = rspApdu.p_data[0];
+      mOpenedchannelCount++;
+      mOpenedChannels[resApduBuff.channelNumber] = true;
+      sestatus = SESTATUS_SUCCESS;
+    } else {
+      resApduBuff.channelNumber = -1;
+      sestatus = ISecureElement::CHANNEL_NOT_AVAILABLE;
+    }
   } else if (((rspApdu.p_data[rspApdu.len - 2] == 0x6E) ||
               (rspApdu.p_data[rspApdu.len - 2] == 0x6D)) &&
              rspApdu.p_data[rspApdu.len - 1] == 0x00) {
-- 
2.55.0

